### DISCUSSION TITLE
Why does npm not give a mechanism for maintainers to recover unpublished versions by mistake?

### DISCUSSION THREAD

**Discussion by aleen42** [npm/rfcs#690] [General]
As we know, if we unpublished a version, the version can no longer publish again, which means that some private repositories which depend on it should break. Can npm give maintainers a mechanism to recover them?

**Comment by ljharb** [comment]
Recover how? Like, un-unpublish?

The version number should never be used for anything *different* at least.

**Reply by aleen42** [reply]
Does NPM clear their database after unpublished immediately? If not, it can give maintainers a chance whithin this period to recover only the same release, but not the same version with new code.

**Reply by ljharb** [reply]
I would assume that un-unpublishing could cause problems with mirrors and private registries, since an unpublish is synonymous with deletion.

What’s the downside to just bumping the patch?

**Reply by aleen42** [reply]
Mirrors or private registries are always synchronous with the central mirror (npm), and the problem may occur if they throw an error when founding an unpublished version has been released again.

The disadvantage of restricting bumping the same version may be:

1. If a maintainer unpublish a version used by private repositories by mistake, which npm cannot scan out, he has broken all those repositories, which should upgrade the version.
2. If a maintainer unpublish a version forcibly like someone before, which has hijacked the dependencies of many repositories, NPM still needs a mechanism to un-unpublish it and protect it.

**Reply by ljharb** [reply]
Publishing a patch will address those issues.

In the case of a security incident I’m sure npm has the capability if you reach out to support - that doesn’t necessarily warrant a self-serve feature.

**Reply by aleen42** [reply]
Can we publish a patch with the same version without breaking other repositories to make the upgrade?

**Reply by ljharb** [reply]
"a patch" means it's not the same version, it's the version with the third number incremented by one - but certainly it can have the same code as the unpublished version.

**Reply by aleen42** [reply]
It works when other repositories specify dependencies with the suffix `~`, or `^`, but if repositories specify a specific version, they need to upgrade to specify the patched version. There is still breaking. I hope that any other repositories depending on this dependency won't be affected by a mistake unpublished.

**Reply by ljharb** [reply]
Yes, that's why it's a bad practice to do that :-)
